zoneinfo/src/types.rs

```diff
@@ -0,0 +1,650 @@
+pub enum RuleIdentifier {
+    None,
+    Numeric(Time),
+    Named(String),
+}
+
+impl TryFromStr<LineParseContext> for RuleIdentifier {
+    type Error = ZoneInfoParseError;
+    fn try_from_str(s: &str, ctx: &mut LineParseContext) -> Result<Self, Self::Error> {
+        ctx.enter("RuleIdentifier");
+        if s == "-" {
+            ctx.exit();
+            return Ok(Self::None);
+        }
+        if s.contains(":") {
+            ctx.exit();
+            return Time::try_from_str(s, ctx).map(Self::Numeric);
+        }
+        ctx.exit();
+        Ok(Self::Named(s.to_owned()))
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum AbbreviationFormat {
+    String(String),
+    Numeric,
+    Pair(String, String),
+    Formattable(FormattableAbbr),
+}
+
+impl AbbreviationFormat {
+    pub(crate) fn format_with_transition(&self, transition: &Transition) -> String {
+        self.format(
+            transition.offset,
+            transition.letter.as_deref(),
+            transition.dst,
+        )
+    }
+
+    pub fn format(&self, offset: i64, letter: Option<&str>, is_dst: bool) -> String {
+        match self {
+            Self::String(s) => s.clone(),
+            Self::Formattable(s) => s.to_formatted_string(letter.unwrap_or("")),
+            Self::Pair(std, dst) => {
+                if is_dst {
+                    dst.clone()
+                } else {
+                    std.clone()
+                }
+            }
+            Self::Numeric => offset_to_str(offset),
+        }
+    }
+}
+
+fn offset_to_str(n: i64) -> String {
+    let mut output = String::new();
+    if n.is_positive() {
+        write!(&mut output, "+").expect("failed to write");
+    } else {
+        write!(&mut output, "-").expect("failed to write");
+    }
+    let hour = n.abs().div_euclid(3600);
+    write!(&mut output, "{hour:02}").expect("failed to write");
+    let minute = n.abs().rem_euclid(3600).div_euclid(60);
+    if minute > 0 {
+        write!(&mut output, "{minute:02}").expect("failed to write");
+    }
+    output
+}
+
+impl TryFromStr<LineParseContext> for AbbreviationFormat {
+    type Error = ZoneInfoParseError;
+    fn try_from_str(s: &str, ctx: &mut LineParseContext) -> Result<Self, Self::Error> {
+        ctx.enter("Abbr. Format");
+        let value = if s.contains("%s") {
+            Ok(Self::Formattable(FormattableAbbr(s.to_owned())))
+        } else if s.contains("%z") {
+            Ok(Self::Numeric)
+        } else if s.contains("/") {
+            let (std, dst) = s.split_once('/').ok_or(ZoneInfoParseError::UnknownValue(
+                ctx.line_number,
+                s.to_owned(),
+            ))?;
+            Ok(Self::Pair(std.to_owned(), dst.to_owned()))
+        } else {
+            Ok(AbbreviationFormat::String(s.to_owned()))
+        };
+        ctx.exit();
+        value
+    }
+}
+
+impl Default for AbbreviationFormat {
+    fn default() -> Self {
+        Self::String("LMT".to_owned())
+    }
+}
```

Comment:
suggestion: It doesn't appear at a glance that this `Default` impl is used and I'm not sure that `AbbreviationFormat` has a clear default value.

Reply:
Yeah, I missed this / forgot to remove it when I made changes from earlier versions.

The concept at the time was that first row values were "default" unless specified (not a good approach)